### ISSUE TITLE
race condition when registering instance ?

### ISSUE THREAD

**Issue by alessnet** [Netflix/eureka#802]
Hello,

I'm using the REST api to register my service in Eureka.
There is a case when one of my Eureka server doesn't see the latest status of my registered service.

Let's say I have two Eureka instances MACHINE1 and MACHINE2.
When my service starts, it takes the first available server (MACHINE1) and does a register (POST /v2/apps/myservice) with the status "STARTING".

The service does start and then register again with the status "UP".
If for any reason (let's say a connection issue) the second call to register is done on MACHINE2 right after the first call, then the service is seen as "STARTING" on MACHINE2 and "Up" on MACHINE1, no matter how long we wait (heartbeats are sent).

In the logs of MACHINE2, I have the following warning:

> 2016-05-31 09:04:38,614 INFO  com.netflix.eureka.registry.AbstractInstanceRegistry:266 [http-nio-8080-exec-44] [register] Registered instance myservice/inst1 with status UP (replication=false)

> 2016-05-31 09:04:39,098 WARN  com.netflix.eureka.registry.AbstractInstanceRegistry:209 [http-nio-8080-exec-3] [register] **There is an existing lease and the existing lease's dirty timestamp 1464710678627 is greater than the one that is being registered 1464710678020**

> 2016-05-31 09:04:39,098 INFO  com.netflix.eureka.registry.AbstractInstanceRegistry:266 [http-nio-8080-exec-3] [register] Registered instance myservice/inst1 with status STARTING (replication=true)

> 2016-05-31 09:06:37,723 INFO  com.netflix.eureka.registry.AbstractInstanceRegistry:330 [http-nio-8080-exec-32] [internalCancel] Cancelled instance myservice/inst1 (replication=false)

If I wait 5 seconds between "Starting" and "Up", the issue disappears.

So I suppose that the second register call is processed BEFORE the replication takes place. When the registration takes place, the "expired" registration is kept anyway (ignoring dirty timestamp). I'm using Eureka 1.4.8.

```
if (existingLease != null && (existingLease.getHolder() != null)) {
                Long existingLastDirtyTimestamp = existingLease.getHolder().getLastDirtyTimestamp();
                Long registrationLastDirtyTimestamp = r.getLastDirtyTimestamp();
                logger.debug("Existing lease found (existing={}, provided={}", existingLastDirtyTimestamp, registrationLastDirtyTimestamp);
                if (existingLastDirtyTimestamp > registrationLastDirtyTimestamp) {
                    logger.warn("There is an existing lease and the existing lease's dirty timestamp {} is " +
                                    "greater than the one that is being registered {}",
                            existingLastDirtyTimestamp,
                            registrationLastDirtyTimestamp);
                    r.setLastDirtyTimestamp(existingLastDirtyTimestamp);
                }
```

Can you please explain if I made something wrong or if indeed Eureka doesn't handle this well ?

Thank you,

Alessandro De Simone

**Comment by qiangdavidliu** [comment]
@alessnet are you using eureka via direct REST requests, and not through the provided eureka-client? There are a few subtleties in the way registrations are handled by eureka, that is usually abstracted away by using the client. Primarily, these are:
- the client contains a eureka specific loadbalancer for talking to the server cluster, such that it minimises target switches for a given client:server pair.
- the client abstract away the usage of the InstanceInfo.lastDirtyTimestamp field in the registration protocol, which is used to provide linearization for registration events even if they end up going to different eureka servers.

**Comment by alessnet** [comment]
@qiangdavidliu

I'm using the REST API in C#. Using the java client (event in a separate process) not an option unfortunately.

I'm specifying the dirtytimestamp myself during registration at client side (in case the timers on the two eureka servers are slightly off). But the issue persists if I don't specify it. What should I put in this field ?

I agree that a correct load balancer will minimize switches and will try to stick to MACHINE1 for the second register. But in my understanding you can still have this issue in real live production. If MACHINE1 doesn't answer to the second register, the load balancer will redirect the call to MACHINE2. If the replication between MACHINE1 and MACHINE2 is slow for some reason, you will end up with the exact same situation where the second register is discarded on MACHINE2 after the replication.

Can you please explain me how the java client handles this situation ? (or lead me more precisely to the source file). 

Thank you ! :)

**Comment by qiangdavidliu** [comment]
@alessnet the lastDirtyTimestamp is intended to be supplied at the client side (and is what the java client does, see [InstanceInfoReplicator.java](https://github.com/Netflix/eureka/blob/master/eureka-client/src/main/java/com/netflix/discovery/InstanceInfoReplicator.java) for code and also comments on client registration logic).

Since this intended to be client generated (and guaranteed at the client side to be serialized), it is possible to use this value as the mechanism to serialize write operations (register, update and cancel). The server side replication logic also utilize the dirtyTimestamp to deal with conflict as you described.

**Comment by alessnet** [comment]
@qiangdavidliu

First of all, thank you for taking the time to answer my questions :)

I'm sending a dirty timestamp and commands are correctly serialized at clientside (so the second register is correctly sent after the first one correctly finished). This is similar to your java code, so I can assure you that I don't make two write operations in parallel.

The only difference I see is that every command is delayed by x seconds while my implementation does not. And that was my original question : if you say that the race condition I had could be avoided only by delaying each command, how is that possible? There always could be a lag/partition in the network that could last longer that the chosen delay. Waiting at client side to hope for a correct sync between replicas seems wrong.

> The server side replication logic also utilize the dirtyTimestamp to deal with conflict as you described.

In the log I have attached, you see that one of the server (which will end up showing the wrong state) print a warning about a old dirtytimestamp (replication=true). If I understand this correctly, in  [AbstractInstanceRegistry](https://github.com/Netflix/eureka/blob/da2f9e5bcf2503487a52d839093f039bea065af2/eureka-core/src/main/java/com/netflix/eureka/registry/AbstractInstanceRegistry.java#L208) the server takes the expired entry ("STARTING") anyway but keep the more recent dirtystimestamp (the one of the "UP"). It seems that If the server had ignored this expired command, the server would have kept the "UP" as expected.

When this occurs, no matter how much time I wait, the two servers doesn't see the same status. Why is this related to the client and why the servers cannot agree of the final status?

**Comment by qiangdavidliu** [comment]
@alessnet sorry about only getting back to you now, I was away for a bit.

I did some more digging into the code, and I think you are correct in saying that (if using the REST api directly), there is a potential registration race condition where if two successive registration requests landed on different machines and happen fast enough such that replication has not yet occurred between the two machines, it may lead to the inconsistency you observed. I believe the provided java client's error handling (inadvertently) guards against this which is why we have not seen it in our production systems.

I'm going to see if I can repro this and work on a fix if it is confirmed. Thanks again for point this out!

**Comment by alessnet** [comment]
Hi @qiangdavidliu, did you manage to reproduce this?

**Comment by qiangdavidliu** [comment]
Hi @alessnet we have not had time to work on this unfortunately. Will see if we can try to repro this this week.

**Comment by qiangdavidliu** [comment]
@alessnet I've semi-verified it via the above PR, and will be doing some tests in our test and prod environments before confirming.

**Comment by qiangdavidliu** [comment]
Fixed in release 1.4.9

**Comment by qiangdavidliu** [comment]
Actually 1.4.9 provided a mechanism to "fix" this, but has not defaulted it yet. Will keep this open until we switch the defaults over.

**Comment by qiangdavidliu** [comment]
Released in 1.4.11
